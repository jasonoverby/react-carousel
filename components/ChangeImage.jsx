import React from 'react';
import PropTypes from 'prop-types';

import { updateImage, getNewImageIndex } from '/lib/utils';

export const DecrementImage = ({
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => (
  <button
    aria-label="show previous image"
    className="changeImage"
    onClick={async () => {
      const updatedImageIndex = getNewImageIndex(imageIndex, 'decrement');
      await updateImage({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    }}
  >
    <span className="arrow">◀</span>
  </button>
);

export const IncrementImage = ({
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => (
  <button
    aria-label="show previous image"
    className="changeImage"
    onClick={async () => {
      const updatedImageIndex = getNewImageIndex(imageIndex, 'increment');
      await updateImage({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    }}
  >
    <span className="arrow">▶</span>
  </button>
);

DecrementImage.propTypes = {
  imageIndex: PropTypes.number,
  setActiveDotIndex: PropTypes.func,
  setImageIndex: PropTypes.func,
  setImageLoaded: PropTypes.func,
};
IncrementImage.propTypes = DecrementImage.propTypes;
