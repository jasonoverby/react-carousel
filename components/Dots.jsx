import React from 'react';

import { imageLocations } from '/lib/image-locations';
import { updateImage } from '/lib/utils';

export const Dots = ({
  activeDotIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  const dots = imageLocations.map((_, i) => (
    <Dot
      index={i}
      activeDotIndex={activeDotIndex}
      setActiveDotIndex={setActiveDotIndex}
      updatedImageIndex={i}
      setImageIndex={setImageIndex}
      setImageLoaded={setImageLoaded}
      key={`dot-${i}`}
    />
  ));

  return <div className="dots">{dots}</div>;
};

export const Dot = ({
  activeDotIndex,
  index,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
  updatedImageIndex,
}) => {
  return (
    <button
      className={index === activeDotIndex ? 'active' : ''}
      // image number in label should not be zero-indexed as this could be confusing
      aria-label={`show image ${index + 1}`}
      onClick={async () => {
        await updateImage({
          setActiveDotIndex,
          setImageIndex,
          setImageLoaded,
          updatedImageIndex,
        });
      }}
    >
      <span>â€¢</span>
    </button>
  );
};
