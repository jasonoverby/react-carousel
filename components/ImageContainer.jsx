import React from 'react';
import PropTypes from 'prop-types';

import { imageLocations } from '../lib/image-locations';

export const ImageLink = ({ imageIndex, imageLoaded, setCarouselPaused }) => {
  return (
    <a
      aria-label={`Visit
        ${imageLocations[imageIndex].href} (opens in a new tab)`}
      href={imageLocations[imageIndex].href}
      onMouseOver={() => setCarouselPaused(true)}
      // for keyboard navigation
      onFocus={() => setCarouselPaused(true)}
      onMouseOut={() => setCarouselPaused(false)}
      // for keyboard navigation
      onBlur={() => setCarouselPaused(false)}
      rel="noreferrer"
      target="_blank"
    >
      <img
        alt={imageLocations[imageIndex].alt}
        aria-label="This image will change every ten seconds unless the cursor is hovering over it"
        src={imageLocations[imageIndex].src}
        className={imageLoaded ? 'imageLoaded' : ''}
      />
    </a>
  );
};

export const ImageContainer = ({
  imageIndex,
  imageLoaded,
  setCarouselPaused,
}) => (
  <div className="imgContainer">
    <ImageLink
      imageIndex={imageIndex}
      imageLoaded={imageLoaded}
      setCarouselPaused={setCarouselPaused}
    />
  </div>
);

ImageContainer.propTypes = {
  imageIndex: PropTypes.number.isRequired,
  imageLoaded: PropTypes.bool.isRequired,
  setCarouselPaused: PropTypes.func.isRequired,
};
ImageLink.propTypes = ImageContainer.propTypes;
