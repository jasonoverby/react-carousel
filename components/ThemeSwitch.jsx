import React from 'react';
import PropTypes from 'prop-types';

export const ThemeSwitch = ({ checked, toggleTheme }) => {
  const handleKeyPress = (e) => {
    // if user presses spacebar when switch is in focus
    // theme is toggled
    if (e.keyCode === 32) {
      // prevents default browser scroll on spacebar behavior
      e.preventDefault();
      toggleTheme();
    }
  };
  return (
    <>
      <span className="themeIcon">ðŸŒž</span>
      <div className="themeSwitch">
        <label tabIndex="0" onKeyDown={handleKeyPress}>
          <input
            type="checkbox"
            role="switch"
            onChange={toggleTheme}
            checked={checked}
            aria-checked={checked}
          />
          <span className="slider"></span>
        </label>
      </div>
      <span className="themeIcon">ðŸŒœ</span>
    </>
  );
};

ThemeSwitch.propTypes = {
  checked: PropTypes.bool,
  toggleTheme: PropTypes.func,
};
