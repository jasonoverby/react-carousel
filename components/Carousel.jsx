import React, { useEffect, useState } from 'react';

import { ImageContainer } from './ImageContainer';
import { Dots } from './Dots';
import { DecrementImage, IncrementImage } from './ChangeImage';
import { getNewImageIndex, updateImage } from '../lib/utils';

export const Carousel = () => {
  const [imageIndex, setImageIndex] = useState(0);
  const [carouselPaused, setCarouselPaused] = useState(false);
  const [imageLoaded, setImageLoaded] = useState(true);
  const [activeDotIndex, setActiveDotIndex] = useState(0);

  useEffect(() => {
    const intervalId = setInterval(async () => {
      if (!carouselPaused) {
        const updatedImageIndex = getNewImageIndex(imageIndex, 'increment');
        await updateImage({
          setActiveDotIndex,
          setImageIndex,
          setImageLoaded,
          updatedImageIndex,
        });
      }
    }, 10000);
    return () => clearInterval(intervalId);
  });

  return (
    <div className="container">
      <DecrementImage
        imageIndex={imageIndex}
        setActiveDotIndex={setActiveDotIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
      <ImageContainer
        imageIndex={imageIndex}
        imageLoaded={imageLoaded}
        setCarouselPaused={setCarouselPaused}
      />
      <IncrementImage
        imageIndex={imageIndex}
        setActiveDotIndex={setActiveDotIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
      <Dots
        activeDotIndex={activeDotIndex}
        setActiveDotIndex={setActiveDotIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
    </div>
  );
};
