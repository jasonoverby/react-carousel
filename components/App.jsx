import React from 'react';

import { Carousel } from './Carousel';

export const App = () => {
  return <Carousel />;
};
