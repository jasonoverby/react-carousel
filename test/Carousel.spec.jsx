import React from 'react';
import { render } from '@testing-library/react';

import { Carousel } from '../components/Carousel';

describe('Carousel', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Carousel />);
    expect(asFragment()).toMatchSnapshot();
  });
});
