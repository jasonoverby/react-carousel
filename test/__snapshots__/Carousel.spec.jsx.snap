// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <button
      aria-label="show previous image"
      class="changeImage"
      type="button"
    >
      <span
        class="arrow"
      >
        ◀
      </span>
    </button>
    <div
      class="imgContainer"
    >
      <a
        aria-label="Visit
        https://www.scientificamerican.com/article/cats-can-get-coronavirus-study-suggests-but-pet-owners-need-not-panic/ (opens in a new tab)"
        href="https://www.scientificamerican.com/article/cats-can-get-coronavirus-study-suggests-but-pet-owners-need-not-panic/"
        rel="noreferrer"
        target="_blank"
      >
        <img
          alt="cat on a bed"
          aria-label="This image will change every ten seconds unless the cursor is hovering over it"
          class="imageLoaded"
          src="https://bit.ly/370FSY3"
        />
      </a>
    </div>
    <button
      aria-label="show next image"
      class="changeImage"
      type="button"
    >
      <span
        class="arrow"
      >
        ▶
      </span>
    </button>
    <div
      class="dots"
    >
      <button
        aria-label="show image 1"
        class="active"
        type="button"
      >
        <span>
          •
        </span>
      </button>
      <button
        aria-label="show image 2"
        class=""
        type="button"
      >
        <span>
          •
        </span>
      </button>
      <button
        aria-label="show image 3"
        class=""
        type="button"
      >
        <span>
          •
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
